engine/src/main/java/org/terasology/rendering/opengl/AbstractFBM.java

```diff
@@ -0,0 +1,269 @@
+ * outline:  greyscale depth-based rendering of object outlines
+ * ssao:  greyscale screen-space ambient occlusion rendering
+ * ssaoBlurred:  greyscale screen-space ambient occlusion rendering - blurred version
+ * scenePrePost:  intermediate step, combining a number of renderings made available so far
+ * lightShafts:  light shafts rendering
+ * sceneHighPass:  a number of buffers to create the bloom effect
+ * sceneBloom0:
+ * sceneBloom1:
+ * sceneBloom2:
+ * sceneBlur0:  a pair of buffers holding blurred versions of the rendered scene,
+ * sceneBlur1:  also used for the bloom effect, but not only.
+ * ocUndistorted:  if OculusRift support is enabled this buffer holds the side-by-side views
+ * for each eye, with no lens distortion applied.
+ * sceneFinal:  the content of this buffer is eventually shown on the display or sent to a file if taking a screenshot
+ */
+
+/**
+ * TODO: fix above, add javadocs
+ */
+public abstract class AbstractFBM implements BaseFBM {
+    protected static final Logger logger = LoggerFactory.getLogger(AbstractFBM.class);
+    protected Map<ResourceUrn, FBOConfig> fboConfigs;
+    protected Map<ResourceUrn, FBO> fboLookup;
+    protected Map<ResourceUrn, Integer> fboUsageCountMap;
+
+
+    private Set<FBOManagerSubscriber> fboManagerSubscribers;
+
+    @Override
+    public void initialise() {
+        fboLookup = Maps.newHashMap();
+        fboUsageCountMap = Maps.newHashMap();
+        fboConfigs = Maps.newHashMap();
+        fboManagerSubscribers = Sets.newHashSet();
+    }
+
+    protected FBO generate(FBOConfig fboConfig, FBO.Dimensions dimensions) {
+        FBO fbo = FBO.create(fboConfig, dimensions);
+
+        // At this stage it's unclear what should be done in this circumstances as I (manu3d) do not know what
+        // the effects of using an incomplete FrameBuffer are. Throw an exception? Live with visual artifacts?
+        if (fbo.getStatus() == FBO.Status.INCOMPLETE) {
+            logger.error("FBO " + fboConfig.getResourceUrn() + " is incomplete. Look earlier in the log for details.");
+        } else if (fbo.getStatus() == FBO.Status.UNEXPECTED) {
+            logger.error("FBO " + fboConfig.getResourceUrn() + " has generated an unexpected status code. Look earlier in the log for details.");
+        }
+        fboLookup.put(fboConfig.getResourceUrn(), fbo);
+        fboConfigs.put(fboConfig.getResourceUrn(), fboConfig);
+        return fbo;
+    }
+
+    protected void notifySubscribers() {
+        for (FBOManagerSubscriber subscriber : fboManagerSubscribers) {
+            subscriber.update();
+        }
+    }
+
+
+    protected void retain(ResourceUrn resourceUrn) {
+        if (!fboUsageCountMap.containsKey(resourceUrn)) {
+            fboUsageCountMap.put(resourceUrn, 1);
+            return;
+        }
+        int usageCount = fboUsageCountMap.get(resourceUrn) + 1;
+        fboUsageCountMap.put(resourceUrn, usageCount);
```

Comment:
You should wrap the last two lines in the IF block and place them first, as the most likely case.